Implementation of recognize_face method in security_system.py

Aqua-16 opened the pull request:
## Description

Imported the `datetime` module into the code.
Detected face using `detectMultiScale` function.
Used the trained model to recognize the face from its database.
Made a unique bounding box around the face. [I do like the frame a lot]
Used the label recognized to get name and authorization status of the person from its database.
Annotated the image with its details above it.
Called the function to log the access attempt into the file.

Fixes #7 : Implemented the function as required.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Motivation and Context

This change was required to implement real-time face recognition feature in the `security_system.py` file. Now, faces can be detected from a video stream in real-time and shown to the user.

## How Has This Been Tested?

I tested this on a few images to check how the text is shown on different frames.

## Checklist:

- [x] I have registered myself at [Contrihub website](https://sac.mnnit.ac.in/contrihub).
- [x] My code follows the code style of this project.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

Aqua-16 committed:
Update security_system.py

Implemented the #TODO 1 sections

Aqua-16 committed:
Update security_system.py (Version 2)

Aqua-16 committed:
Update security_system.py (Version 3)

Fixed the error of wrong indentation in loading the model. 
The wrong indentation would have not loaded the model if a different model_path had been provided.

Aqua-16 committed:
Merge branch 'ContriHUB:main' into main

Aqua-16 committed:
Implemented recognize_face method

Perform Face Detection and Recognition
Draw unique bounding box around face
Annotate image with details
Add log of attempt

Aqua-16 committed:
Refactor: Logging implementation

Changed how I implemented logging access attempt to use another function instead of writing code for logging again.

sanskaromar commented on app/security_system.py:
```diff
@@ -66,13 +67,36 @@ def recognize_face(self, frame):
+            # Extract the detected face region
+            x,y,w,h = face
+            X,Y = x+w,y+h
+            image = gray_frame[x:X, y:Y]
```
Sus.

sanskaromar commented on app/security_system.py:
```diff
@@ -66,13 +67,36 @@ def recognize_face(self, frame):
+            cv2.line(frame,(X-l,Y),(X,Y),(255,0,200),t)
+            cv2.line(frame,(X,Y-l),(X,Y),(255,0,200),t)
+            # Annotate the frame with the recognized user's name and authorization status
+            name = self.get_person_name(label,confidence)
```
Variable naming should be meaningful. `name` could be `person_name`.

sanskaromar commented on app/security_system.py:
```diff
@@ -66,13 +67,36 @@ def recognize_face(self, frame):
+            cv2.line(frame,(X,Y-l),(X,Y),(255,0,200),t)
+            # Annotate the frame with the recognized user's name and authorization status
+            name = self.get_person_name(label,confidence)
+            authorization_status = self.authorized_persons[name]
```
The `authorization_status` will hold values `true` or `false`. So, it makes more sense to put the name of variable as something like `is_authorized`.

sanskaromar commented on app/security_system.py:
```diff
@@ -66,13 +67,36 @@ def recognize_face(self, frame):
+            # Draw bounding box around the detected face
+            l= 20
+            t= 5
+            cv2.rectangle(frame,face,(255,0,200),1)
```
Can we draw bounding boxes with different colors for different authorization statuses? For example: Green if authorized and red if not.

sanskaromar reviewed (changes requested):
Please make the requested changes. I have not mentioned one change you can make to reduce your code further. Let's see if you catch that, else I'll tell you in the next review.

Also, don't forget to submit PR request in Contrihub to get your points.

Aqua-16 commented:
Accidentally made it for review.

Aqua-16 committed:
Feat color change, fix image selection, minor refactors

Refactored authorization_status values passed to logger.
Fixed image extraction from grayframe. 
Implemented feature for bounding box to change color based on authorization status.

Aqua-16 commented:
I have streamlined the code a bit, but I wasn't able to find any other way to shorten the code.

sanskaromar commented on app/security_system.py:
```diff
@@ -66,13 +67,43 @@ def recognize_face(self, frame):
+            if(authorization_status==True):
+                color = (0,200,100)
+                is_authorized = "YES"
+            else:
+                color = (0,35,200)
+                is_authorized = "NO"
```
Prefer red color for not authorized.

Aqua-16 replied:
Sir I did prefer red for unauthorized.
![image](https://github.com/ContriHUB/SecureFace/assets/123102778/574807c3-44ba-42ff-ae10-31b75c66e2fe)

sanskaromar replied:
Oh My Bad! I read it as RGB instead of BGR. 👍🏻

sanskaromar commented on app/security_system.py:
```diff
@@ -66,13 +67,43 @@ def recognize_face(self, frame):
+            # Perform facial recognition on the detected face
+            label, confidence = self.face_recognizer.predict(image)
+            person_name = self.get_person_name(label,confidence)
+            authorization_status = self.authorized_persons[person_name]
+            if(authorization_status==True):
+                color = (0,200,100)
+                is_authorized = "YES"
+            else:
+                color = (0,35,200)
+                is_authorized = "NO"
```
Try to use the already implemented function `is_person_authorized` to get the authorization status. We're doing this because currently we have a simple file to get the authorization status but later we might wanna change that, in that case it makes more sense to implement it as a separate function.

You might need to change the function body of `is_person_authorized` as it may not be correctly implemented.

sanskaromar commented on app/security_system.py:
```diff
@@ -66,13 +67,43 @@ def recognize_face(self, frame):
+            cv2.putText(frame,f"Name: {person_name}",(x+5,y-20),cv2.FONT_HERSHEY_PLAIN,0.8,(30,30,30),1)
+            cv2.putText(frame,f"Authorization Status: {authorization_status}",(x+5,y-10),cv2.FONT_HERSHEY_PLAIN,0.8,(0,0,0),1)
+            # Log access attempt
+            self.log_access_attempt(person_name,is_authorized)
```
Here are our sample logs:
```
2023-10-05 19:06:10,989 - Person: Unknown, Authorized: False
2023-10-05 19:06:11,055 - Person: Madonna, Authorized: True
2023-10-05 19:06:11,120 - Person: Elton John, Authorized: True
```

We want to put True/False instead of Yes/No there.

sanskaromar commented:
Also, pull the changes with logging implemented to understand why I'm stressing upon `is_Authorized` being `True/False`.

Aqua-16 committed:
Merge branch 'ContriHUB:main' into main

Aqua-16 committed:
Refactored is_authorized, Used function for authorization_status

Changed is_authorized values from Yes/No to True/False
Hard coded value of text rectangle to prevent image size from affecting it.
called is_person_authorized() method to set authorization_status for scalability.

Additionally, pulled changes from logging.

sanskaromar commented on app/security_system.py:
```diff
@@ -72,13 +73,43 @@ def recognize_face(self, frame):
+                is_authorized = "False"
+            # Draw bounding box around the detected face
+            l= 20
+            t= 5
```
Variable `thickness` would make code readable than `t`.

Aqua-16 replied:
That is true. I'll edit that in my next commit.

sanskaromar reviewed (approved):
Great Work @Aqua-16 🚀
